compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1733,11 +1738,28 @@ impl<'a> Parser<'a> {
             self.sess.gated_spans.gate(sym::async_closure, span);
         }
 
-        Ok(self.mk_expr(
+        if self.token.kind == TokenKind::Semi && self.token_cursor.frame.delim == DelimToken::Paren
+        {
+            // It is likely that the closure body is a block but where the
+            // braces have been removed. We will recover and eat the next
+            // statements later in the parsing process.
+            body = self.mk_expr_err(body.span);
+        }
```

Comment:
I'm somewhat uneasy with this check, but having both `Semi` and `Paren` be true should only happen on malformed code.